### ISSUE TITLE
```torch.distributions.Categorical``` unintended  ```log_prob``` gradient w.r.t ```probs```

### ISSUE THREAD

**Issue by TolgaOk** [pytorch/pytorch#61727]
## 🐛 Bug

When ```torch.distributions.Categorical``` is initialized with ```probs```, the implementation normalizes it even if it is already normalized. However, if we give normalized values to ```probs```, this normalization leads to incorrect gradients. This happens when we have the softmax output of some tensor and we want to use Categorical distribution with the normalized prob values that we have.

## To Reproduce

```Python
import torch

normalized_probs = torch.tensor([[0.2, 0.8]], dtype=torch.float32, requires_grad=True)

dist = torch.distributions.Categorical(probs=normalized_probs)
sample = dist.sample()
dist.log_prob(sample).sum().backward()

print(normalized_probs.grad, sample)
>>> tensor([[ 4., -1.]]) tensor([0])
```
We expect to see zero in the gradient at index 1 since that index is not sampled, but we observed -1.0

## Expected behavior

The correct gradient can be obtained by ignoring the additional normalization performed by ```torch.distributions.Categorical``` at the initialization.

```Python
normalized_probs = torch.tensor([[0.2, 0.8]], dtype=torch.float32, requires_grad=True)

dist = torch.distributions.Categorical(probs=normalized_probs)
dist.probs = normalized_probs  # This line solves the issue
dist.log_prob(sample).sum().backward()

print(normalized_probs.grad, sample)
>>> tensor([[5., 0.]]) tensor([0])
```
This time the gradient at index 1 is zero.

## Environment

PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 10.15.7 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.27)
CMake version: version 3.19.0
Libc version: N/A

Python version: 3.8.3 (default, Jul  2 2020, 11:26:31)  [Clang 10.0.0 ] (64-bit runtime)
Python platform: macOS-10.15.7-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] numpydoc==1.1.0
[pip3] torch==1.7.0
[pip3] torchaudio==0.7.0a0+ac17b64
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.8.1
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      233
[conda] mkl-service               2.3.0            py38hfbe908c_0
[conda] mkl_fft                   1.1.0            py38hc64f4ea_0
[conda] mkl_random                1.1.1            py38h959d312_0
[conda] numpy                     1.18.5           py38h1da2735_0
[conda] numpy-base                1.18.5           py38h3304bdc_0
[conda] numpydoc                  1.1.0                      py_0
[conda] pytorch                   1.7.0                   py3.8_0    pytorch
[conda] torchaudio                0.7.0                      py38    pytorch
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.8.1                  py38_cpu    pytorch

**Comment by fritzo** [comment]
Note the two gradients are equivalent modulo a constant shift.  I'm not sure whether there is a canonical representation of gradients in a submanifold, such as here the submanifold of normalized probability vectors.  I guess we've assumed so far that we can generate any gradient that projects down to some canonical tangent gradient.  Does anyone know a standard mathematical solution?

@TolgaOk it is not clear to me that this is an actual bug, since the two gradients are equivalent.  Can you provide an example where an unconstrained variable is transformed to the constrained space of probs, and where the gradients on that unconstrained variable differ between the two implementations of `Categorical.log_prob()`?  When I try to compute grads in an end-to-end example they are identical up to numerical precision:
```py
import torch
from torch.autograd import grad
from torch.distributions import constraints, biject_to, Categorical
torch.manual_seed(0)

dim = 5
u = torch.randn(dim - 1).requires_grad_()  # an unconstrained parameter we can optimize
t = biject_to(constraints.simplex)         # an injection into the higher dimensional space
p = t(u)                                   # a constrained vector
assert len(p) == len(u) + 1
data = torch.tensor(0)

d = Categorical(p)
print(grad(d.log_prob(data).sum(), [u], retain_graph=True)[0])
>>> tensor([ 4.6140e-01, -4.7519e-09,  0.0000e+00,  0.0000e+00])

d = Categorical(p)
d.probs = p                                # the proposed tweak
print(grad(d.log_prob(data).sum(), [u])[0])
>>> tensor([0.4614, 0.0000, 0.0000, 0.0000])
```

**Comment by TolgaOk** [comment]
Thanks for your comment, @fritzo. I was not sure to call this a bug since this has no significant effect . There is no issue with the unconstrained variable. But for the constrained one, the gradient seems to differ (when we take the gradient w.r.t ```p``` instead of ```u```) by a constant due to additional normalization in the initialization.
To give a perspective, I was visualizing the gradient of a probability vector when I observed this.

**Comment by zcaicaros** [comment]
Hello, this remains till today. I use PyTorch 1.9.0 and can still reproduce it with the following code:
```
import numpy as np
import torch
torch.manual_seed(1)

nn = torch.nn.Linear(2, 2)
input = torch.rand(1, 2)
output = nn(input)

probs = torch.softmax(output, dim=1)
probs.retain_grad()
dis = torch.distributions.categorical.Categorical(probs=probs)
sam = dis.sample()
p = torch.gather(probs, -1, sam.unsqueeze(-1))
print(p)
print(probs)
print(sam)
# log_p = torch.log(p).squeeze()
log_p = dis.log_prob(sam)
# print(log_p)
nn.zero_grad()
log_p.backward()
# print([param.grad for param in nn.parameters()])
print(probs.grad)
```
 Why these two gradients are the same?
